model/model_training/custom_datasets/prompt_dialogue.py

```diff
@@ -193,3 +194,61 @@ def __getitem__(self, idx):
         ]
 
         return DatasetEntrySft(conversation=conv_utt, system_message=instruction)
+
+
+class DolphinMix(Dataset):
+    name = "dophin-mix"
+
+    def __init__(self, cache_dir, num_samples=100000, max_char_len=8000, seed=42):
+        self.dataset = load_dataset(
+            "ehartford/dolphin", data_files="flan5m-alpaca-uncensored.jsonl", cache_dir=cache_dir
+        )
+        self.dataset = self.dataset["train"].shuffle(seed).select(range(num_samples))
+        self.max_char_len = max_char_len
+        instructions = set([item["instruction"] for item in self.dataset])
+
+        self.conversations = []
+        for inst in instructions:
+            data_sample = self.dataset.filter(lambda example: example["instruction"] == inst)
+            available_indices = np.arange(0, len(data_sample)).tolist()
+            removed_indices = []
+            for idx in available_indices:
+                conversation_len = len(inst)
+                if idx not in removed_indices and conversation_len < self.max_char_len:
+                    conversation = {"conversation": []}
+                    conversation["instruction"] = inst
+                    input, output = [data_sample[idx][key] for key in ("input", "output")]
+                    conversation["conversation"].append({"input": input, "output": output})
+                    conversation_len += len(input) + len(output)
+                    removed_indices.append(idx)
+                    while conversation_len < self.max_char_len:
+                        indices_to_pick = np.setdiff1d(available_indices, removed_indices)
```

Comment:
Generating the indices to pick set every time seems a bit costly .. but since it is only done during startup it might be ok.